guice-extension/src/main/java/name/falgout/jeffrey/testing/junit/guice/GuiceExtension.java

```diff
@@ -79,10 +84,44 @@ private static Injector createInjector(ExtensionContext context)
       InvocationTargetException {
     Optional<Injector> parentInjector = getParentInjector(context);
     List<? extends Module> modules = getNewModules(context);
-
+    if (isSharedInjector(context)) {
+      return createCachedInjector(modules, parentInjector);
+    }
     return parentInjector
-        .map(injector -> injector.createChildInjector(modules))
-        .orElseGet(() -> Guice.createInjector(modules));
+            .map(injector -> injector.createChildInjector(modules))
+            .orElseGet(() -> Guice.createInjector(modules));
+  }
+
+  private static Injector createCachedInjector(List<? extends Module> modules,
+      Optional<Injector> parentInjector) {
+    Set<? extends Class<?>> modulClass = modules.stream().map(Object::getClass).collect(toSet());
+    Optional<Injector> cachedInjector = getCachedInjector(modulClass);
+    if (cachedInjector.isPresent()) {
+      return cachedInjector.get();
+    } else {
+      Injector createdInjector = parentInjector
+          .map(injector -> injector.createChildInjector(modules))
+          .orElseGet(() -> Guice.createInjector(modules));
+      if (!modulClass.isEmpty()) {
+        INJECTOR_CACHE.put(modulClass, createdInjector);
```

Comment:
I think there's a logic bug in here. We're only caching the injectors based on the new modules introduced by the current context element.

So we could have something like

```
@IncludeModule(OuterClassModule.class)
class OuterClass {
  @IncludeModule(TestCaseModule.class)
  void outerTestCase() {}

  @IncludeModule(InnerClassModule.class)
  class InnerClass {
    @IncludeModule(TestCaseModule.class)
    void innerTestcase() {}
  }
}
```

and `outerTestCase` and `innerTestCase` would get the same cached injector. And depending on execution order, it may or may not contain `InnerClassModule`.

Could you add a test case for this situation, and make sure that the caching works as expected?

Reply:
For that matter, what happens in the following snippet?

```
@IncludeModule(OuterClassModule.class)
@SharedInjectors
class OuterClass {
  @IncludeModule(ExtraModule.class)
  void testCase1() {}

  @IncludeModule(ExtraModule.class)
  void testCase2() {}
}
```

Do `testCase1` and `testCase2` share an injector? Could you add a test to clarify the behavior here?

Reply:
Inded there was logic bug. Now instead of using name.falgout.jeffrey.testing.junit.guice.GuiceExtension#getNewModules I started using name.falgout.jeffrey.testing.junit.guice.GuiceExtension#getContextModuleTypes(org.junit.jupiter.api.extension.ExtensionContext)  and added tests for this case.